x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java

```diff
@@ -0,0 +1,224 @@
+import org.elasticsearch.compute.operator.EvalOperator.ExpressionEvaluator;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.Expressions;
+import org.elasticsearch.xpack.esql.core.expression.function.OptionalArgument;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.core.type.DataTypes;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamOutput;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.THIRD;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isIPAndExact;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.esql.core.type.DataTypes.INTEGER;
+
+/**
+ * Truncates an IP value to a given prefix length.
+ */
+public class IpPrefix extends EsqlScalarFunction implements OptionalArgument {
+    // Borrowed from Lucene, rfc4291 prefix
+    private static final byte[] IPV4_PREFIX = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, -1 };
+
+    private final Expression ipField;
+    private final Expression prefixLengthV4Field;
+    private final Expression prefixLengthV6Field;
+
+    @FunctionInfo(
+        returnType = "ip",
+        description = "Returns the original IP with all but the first N bits set to zero.",
+        examples = @Example(file = "ip", tag = "ipPrefix")
+    )
+    public IpPrefix(
+        Source source,
+        @Param(
+            name = "ip",
+            type = { "ip" },
+            description = "IP address of type `ip` (both IPv4 and IPv6 are supported)."
+        ) Expression ipField,
+        @Param(
+            name = "prefixLengthV4",
+            type = { "integer" },
+            description = "Prefix length, in the range [0, 32], to apply to IPv4 addresses."
+        ) Expression prefixLengthV4Field,
+        @Param(
+            name = "prefixLengthV6",
+            type = { "integer" },
+            optional = true,
+            description = "Prefix length, in the range [0, 128], to apply to IPv6 addresses. "
+                + "If not provided, the original IPv6 addresses will be returned."
```

Comment:
I'd make it required for now. I don't know what the right behavior should be if it's not provided, but I think skipping ipv6 is probably not it. We can always make that behavior later if this is required.